src/account/balanceHistoryCache.js

```diff
@@ -0,0 +1,135 @@
+// @flow
+
+import { BigNumber } from "bignumber.js";
+import type {
+  BalanceHistoryCache,
+  BalanceHistoryDataCache,
+  AccountLike,
+  Account,
+  SubAccount,
+  GranularityId,
+} from "../types";
+import { getOperationAmountNumberWithInternals } from "../operation";
+import {
+  weekIncrement,
+  dayIncrement,
+  hourIncrement,
+  startOfWeek,
+  startOfDay,
+  startOfHour,
+} from "../portfolio/range";
+
+export const emptyHistoryCache = {
+  HOUR: {
+    latestDate: null,
+    balances: [],
+  },
+  DAY: {
+    latestDate: null,
+    balances: [],
+  },
+  WEEK: {
+    latestDate: null,
+    balances: [],
+  },
+};
+
+const conf = {
+  WEEK: {
+    increment: weekIncrement,
+    startOf: startOfWeek,
+  },
+  DAY: {
+    increment: dayIncrement,
+    startOf: startOfDay,
+  },
+  HOUR: {
+    increment: hourIncrement,
+    startOf: startOfHour,
+  },
+};
+
+function generateHistoryFromOperationsG(
+  account: AccountLike,
+  g: GranularityId
+): BalanceHistoryDataCache {
+  const { increment, startOf } = conf[g];
+  const latestDate = startOf(new Date());
+  const balances = [];
+  let { balance } = account;
+  const operationsLength = account.operations.length;
+  let date = latestDate.getTime();
+  for (let i = 0; i < operationsLength; ) {
+    // accumulate operations after time t
+    while (i < operationsLength && account.operations[i].date > date) {
+      balance = balance.minus(
```

Comment:
Why we compute with BigNumber methods here instead of convert to number immediately?

Reply:
ah, sorry. it's a crypto balance not a countervalue 😅